compiler/rustc_const_eval/src/check_consts/check.rs

```diff
@@ -264,23 +268,47 @@ impl<'mir, 'tcx> Checker<'mir, 'tcx> {
         }
     }
 
-    fn local_has_storage_dead(&mut self, local: Local) -> bool {
+    fn local_is_transient(&mut self, local: Local) -> bool {
         let ccx = self.ccx;
-        self.local_has_storage_dead
+        self.transient_locals
             .get_or_insert_with(|| {
-                struct StorageDeads {
-                    locals: BitSet<Local>,
+                // A local is "transient" if it is guaranteed dead at all `Return`.
+                // So first compute the say of "maybe live" locals at each program point.
+                let always_live_locals = &always_storage_live_locals(&ccx.body);
+                let maybe_storage_live = MaybeStorageLive::new(Cow::Borrowed(always_live_locals))
+                    .into_engine(ccx.tcx, &ccx.body)
+                    .iterate_to_fixpoint()
+                    .into_results_cursor(&ccx.body);
+
+                // And then check all `Return` in the MIR, and if a local is "maybe live" at a
+                // `Return` then it is definitely not transient.
+                struct TransientLocalVisitor<'a, 'tcx> {
+                    maybe_storage_live: ResultsCursor<'a, 'tcx, MaybeStorageLive<'a>>,
+                    transient: BitSet<Local>,
                 }
-                impl<'tcx> Visitor<'tcx> for StorageDeads {
-                    fn visit_statement(&mut self, stmt: &Statement<'tcx>, _: Location) {
-                        if let StatementKind::StorageDead(l) = stmt.kind {
-                            self.locals.insert(l);
+                impl<'a, 'tcx> Visitor<'tcx> for TransientLocalVisitor<'a, 'tcx> {
+                    fn visit_terminator(
+                        &mut self,
+                        terminator: &Terminator<'tcx>,
+                        location: Location,
+                    ) {
+                        if matches!(terminator.kind, TerminatorKind::Return) {
+                            self.maybe_storage_live.seek_after_primary_effect(location);
+                            for local in self.maybe_storage_live.get().iter() {
+                                // If a local may be live here, it is definitely not transient.
+                                self.transient.remove(local);
+                            }
                         }
                     }
                 }
-                let mut v = StorageDeads { locals: BitSet::new_empty(ccx.body.local_decls.len()) };
-                v.visit_body(ccx.body);
-                v.locals
+
+                let mut v = TransientLocalVisitor {
+                    maybe_storage_live,
+                    transient: BitSet::new_filled(ccx.body.local_decls.len()),
+                };
+                v.visit_body(&ccx.body);
```

Comment:
A loop over basic blocks may be simpler than a visitor here. 

You may need to filter out unreachable blocks, as the data flow framework ices when seeking inside those.

Reply:
Ah, is that why [this code](https://github.com/rust-lang/rust/blob/84ac80f1921afc243d71fd0caaa4f2838c294102/compiler/rustc_mir_transform/src/lint.rs#L39-L41) only iterates over `traversal::reachable` blocks?

Is there any way to ask the data flow cursor which blocks are live? It seems quite wasteful to re-compute that information using `traversal::reachable`.

Reply:
> Is there any way to ask the data flow cursor which blocks are live?

Seems like improbable, the `ResultCursor` has a field `reachable_blocks` [here](https://github.com/Jaic1/rust/blob/c8b14ba7b6777dfbfb64a02d7bbe867a54dddf2a/compiler/rustc_mir_dataflow/src/framework/cursor.rs#L34) but only enabled with `#[cfg(debug_assertions)]`.

> It seems quite wasteful to re-compute that information using `traversal::reachable`.

It seems just fine to use `traversal::reachable` to do a forward computation once again, considering the dataflow engine might already iterated the blocks multiple times.

Reply:
All right, I made this do a loop over `traversal::reachable` rather than a visitor.